Add support to configure ocid storage

rhatdan opened the pull request:
Implement ocid-storage-setup.service which executes container-storage-setup
to configure COW Storage.

If you use this script you are requiring container-storage-setup to be
installed.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

rhatdan commented:
This pull needs @rhvgoyal approval before being merged.

mrunalp commented:
LGTM

runcom commented:
LGTM

rhvgoyal commented on contrib/systemd/ocid-storage-setup-override.conf:
```diff
@@ -0,0 +1,7 @@
+# Edit this file to override any configuration options specified in
+# /usr/lib/container-storage-setup/container-storage-setup.
```
Should be /usr/share/ instead of /usr/lib/?

rhatdan replied:
Fixed

rhvgoyal commented on contrib/systemd/ocid-storage-setup-override.conf:
```diff
@@ -0,0 +1,7 @@
+# /usr/lib/container-storage-setup/container-storage-setup.
+#
+# For more details refer to "man container-storage-setup"
+CONFIG_NAME="ocid"
```
This option is yet to be added. We are planning to add it soon.

rhatdan replied:
Yes we will wait to merge until this gets into container-storage-setup

rhvgoyal commented:
container-storage-setup is still version 0.1.0 and some active persistent metadata work is going on. So one can use it for development and testing but we might not support full backward compatibility yet. Once we reach version 1.0.0, we will support full backward compatibility.

mrunalp commented:
@rhvgoyal Does this look good to you?

rhvgoyal commented on contrib/systemd/ocid.service:
```diff
@@ -9,7 +9,7 @@ EnvironmentFile=-/etc/sysconfig/ocid-storage
 Environment=GOTRACEBACK=crash
 ExecStart=/usr/local/bin/ocid \
-          $OCID_STORAGE_OPTIONS \
+          $STORAGE_OPTIONS \
```
@rhatdan Do we also need a `Require=ocid-storage-setup` to make sure ocid-storage-setup gets enabled and run before this service runs. Also Require should make sure if ocid-storage-setup fails, this service fails too.?

rhatdan commented:
I rebased this pull request to handle the name change.  @rhvgoyal I think we should think about changing container-storage-setup to do nothing with the driver if it is not specified.  

In container/storage we have added a config file which allows you to specify where containers are going to be stored on your system.  We specify the default for container/storage there, and skopeo, buildah and crio should grab the default from there.  If a user specifies an alternate driver or location then the tools will nto be able to share storage.

rhvgoyal commented on contrib/rpm/crio.spec:
```diff
@@ -21,6 +21,7 @@ Source0:        https://%{provider_prefix}/archive/%{commit}/%{repo}-%{shortcomm
 Provides:       %{repo}
 
 BuildRequires:  golang-github-cpuguy83-go-md2man
+Requires: container-storage-setup >= 0.3
```
Now we have new version of container-storage-setup (0.6) and that does not conflict with older versions. It might be a good idea to bump up the minimum version requirement to 0.6 instead.

rhvgoyal commented:
Except that minor nit about minimum version requirement, this patch LGTM.

rhvgoyal commented:
@rhatdan, so you are talking about the case of STORAGE_DRIVER="". So what should this tool do in that case? Right now, it does not seem to return an error. Can't remember why am I not giving error if STORAGE_DRIVER is nil. 

Can you explain the use case little more. Why would sombody specify STORAGE_DRIVER=""

rhatdan commented:
Right, if STORAGE_DRIVER is not specified, then it goes with the default storage.  In this case overlay. 
I see container-storage-setup as been a tool for setting up mountpoints for the /var/lib/containers/storage going forward.  I worry a little about us breaking the storage sharing between, skopeo, CRI-O, Kpod, buildah.  None of these other then CRI-O will use containers-storage-setup.

rhvgoyal commented:
Looks like right now STORAGE_DRIVER="" is a nop. I might have done this to allow people to make use of loop devices if they wanted to. (As old code allowed that).

rhvgoyal commented:
I am assuming that all command line based tools work based on path in the filesystem hierarchy. If, that's the case, it should be left to admin/user to setup the storage first properly by running container-storage-setup. (Similar to setting up logical volumes in lvm). And then rest of the tools will work on those mount points.

One problem still remains though. How does storage activate by itself over a reboot. Will it be a good idea to introduce another option where we associate another property with storage configuration. Say "auto-activate". That means container-storage-setup can drop some systemd unit file which runs at boot and activates storage configuration. So by the time tools run, they should find storage ready.

I guess this will be something close to .mount unit files generated by systemd for entries in /etc/fstab.

rhatdan commented:
Yes I think  having the systemd mount service is critical.  Lets talk about this later.

rhatdan commented:
@rhvgoyal and I talked about this, and have decided to close it.  We now believe that we should just have admin setup their storage themselves by editing /etc/containers/storage.conf.  We will write documentation on how to do this.  The problem with containers-storage-setup is we end up with multiple config files.
